README.md
# MotorCortex Subtitles

[Subtitle](https://www.npmjs.com/package/subtitle) library as a MotorCortex Incident

## Installation
```bash
$ npm install @donkeyclip/motorcortex-subtitles
# OR
$ yarn add @donkeyclip/motorcortex-subtitles
```

```javascript
import Subtitle from "@donkeyclip/motorcortex-subtitles";
```


## Key Concepts / Features
MotorCortex Subtitles takes the capabilities of Subtitle library of parsing subtilte files. 
The library exposes just one Incident with the name "ParseText" which can parse supported subtitle files and add them to your clip. The duration is auto calculated from your subtitles. The container params of the subtitles clip should be the same with the parent clip.

## Subtitle Formats supported
SRT or WebVTT

## Documentation
### Import and load the plugin to MotorCortex
```javascript
import { HTMLClip, loadPlugin } from "@donkeyclip/motorcortex/";
import mySubsTextFile from "./subs.srt.js";
import SubtitlesDefinition from "@donkeyclip/motorcortex-subtitles/";
const Subtitles = loadPlugin(SubtitlesDefinition);
const clip = new HTMLClip({
  html: `
    <div class="container"></div>`,
  css: `
  .container{
    width:100%;
    height:100%;
    position:relative;
    background:#151515;
  }
`,
  host: document.getElementById("clip"),
  containerParams: {
    width: "720px",
    height: "640px",
  },
});

const subtitle = new Subtitles.ParseText(
  {
    fontSize: 14,
    textColor: "white",
    fontFamily: "Ubuntu",
    subtitles: mySubsTextFile,
    position: "bottom",
    maxWidth: 400,
    paddingBottom: 50,
  },
  {
    selector: ".container",
    containerParams: {
      width: "720px",
      height: "640px",
    },
  }
);

clip.addIncident(subtitle, 0);
clip.play();
```
### Attributes
| Name | Description | Default | Type |
| --------- |:-----------| :----| ------: |
| fontSize | Font size in pixels of the subtitles | 12 | number |
| textColor | The color of the subtitles | white | - |
| fontFamily | Font family of the subtiltes | 'Ubuntu' | string |
| subtitles | The subtitles loaded text | - | string |
| position | Position of the subtitles container [top, center, bottom] | bottom | string |
| maxWidth | The maximum width of the subtitles container in pixels | "100%" | number |
| paddingTop | The padding top value of the subtitles container in pixels | 0 | number |
| paddingBottom | The padding bottom value of the subtitles container in pixels | 0 | number |

### Demo
https://donkeyclip.github.io/motorcortex-subtitles/demo/


## License
[MIT License](https://opensource.org/licenses/MIT)

  
  
[![Kiss My Button](https://presskit.kissmybutton.gr/logos/kissmybutton-logo-small.png)](https://kissmybutton.gr)


CHANGELOG.md
# Changelog

### 3.0.3 (2021-04-12)

### Features

- add subtitle library ([623f720](https://www.github.com/donkeyclip/motorcortex-subtitles/commit/623f720284fd796173f17aca8345a9bd0dbc1314))
- add subtitle library to handle subtitle files ([f360bf3](https://www.github.com/donkeyclip/motorcortex-subtitles/commit/f360bf3477237944243bca576d99617dcf1cc1d1))
- add validation rules ([3bc629f](https://www.github.com/donkeyclip/motorcortex-subtitles/commit/3bc629f9f936f57c62d69df9c873b1b506280976))
- asdf ([f7fb1ca](https://www.github.com/donkeyclip/motorcortex-subtitles/commit/f7fb1ca026b1f8b5c81d4933435efab3ac57e557))
- html clip implementation ([5be2e25](https://www.github.com/donkeyclip/motorcortex-subtitles/commit/5be2e25a6080eed189a0f23d2e5725ae28e112b7))
- initial working plugin ([5a743a0](https://www.github.com/donkeyclip/motorcortex-subtitles/commit/5a743a0ac0605a65970376593781af671afe9ee1))

### Bug Fixes

- **deps:** update dependency fs to v0.0.2 ([29061fa](https://www.github.com/donkeyclip/motorcortex-subtitles/commit/29061faa8b74e5f80253335ee612dc8c81d2d0cc))
- fix hide subs ([acc5256](https://www.github.com/donkeyclip/motorcortex-subtitles/commit/acc5256ed11d44af867ae905a5655844ea14f88a))
- upate packages ([7a74a69](https://www.github.com/donkeyclip/motorcortex-subtitles/commit/7a74a69bbcee7d47da3168935f9dbb65a2394859))
- update package json ([00d9fb0](https://www.github.com/donkeyclip/motorcortex-subtitles/commit/00d9fb0b5ebacb73d2f07d653cb5140a6fcda3d5))


src/SRT.js
import MC from "@kissmybutton/motorcortex";

export default class SRT extends MC.Effect {
  onGetContext() {
    this.element.style = this.attrs.attrs.css;
    this.subs = this.attrs.animatedAttrs.text
      .trim()
      .split(/\n\n/)
      .map((subtitle) => {
        const splitted = subtitle.split("\n");
        const index = Number(splitted[0]);

        const time = splitted[1].trim();
        const start = time.split("-->")[0];
        const end = time.split("-->")[1];

        const startTime =
          start.split(":")[0] * 3600 +
          start.split(":")[1] * 60 +
          Number(start.split(":")[2].replace(",", ".")) * 1000;
        const endTime =
          end.split(":")[0] * 3600 +
          end.split(":")[1] * 60 +
          Number(end.split(":")[2].replace(",", ".")) * 1000;

        const text = splitted.slice(2, splitted.length).join(" ").trim();

        return {
          index,
          startTime,
          endTime,
          text,
        };
      });
  }

  getScratchValue() {
    return "";
  }
  onProgress(fraction, currentTime) {
    if (fraction == 0) {
      return false;
    }
    for (const i in this.subs) {
      const { startTime, endTime, text } = this.subs[i];
      if (currentTime >= startTime && currentTime < endTime) {
        this.element.innerHTML = text;
        break;
      } else {
        this.element.innerHTML = "";
      }
    }
  }
}


rollup.config.js
import babel from "@rollup/plugin-babel";
import commonjs from "@rollup/plugin-commonjs";
import json from "@rollup/plugin-json";
import resolve from "@rollup/plugin-node-resolve";
import nodePolyfills from "rollup-plugin-polyfill-node";
import { terser } from "rollup-plugin-terser";
import pkg from "./package.json";

export default [
  {
    input: "src/index.js",
    external: ["@donkeyclip/motorcortex"],
    output: [
      {
        file: pkg.main,
        format: "cjs",
      },
      {
        file: pkg.module,
        format: "es",
      },
    ],
    plugins: [
      json(),
      resolve({
        preferBuiltins: false,
      }),
      commonjs(),
      nodePolyfills({ include: "node_modules/**" }),
      babel({}),
    ],
  },
  {
    input: "src/index.js",
    external: ["@donkeyclip/motorcortex"],
    output: [
      {
        globals: {
          "@donkeyclip/motorcortex": "MotorCortex",
        },
        name: pkg.name,
        file: pkg.browser,
        format: "umd",
      },
    ],
    plugins: [
      json(),
      commonjs(),
      nodePolyfills(),
      resolve({ mainFields: ["module", "main", "browser"] }),
      babel(),
      terser(),
    ],
  },
];


src/ParseText.js
import { Effect, HTMLClip, loadPlugin } from "@donkeyclip/motorcortex";
import { parseSync } from "subtitle";

/*INNER PLUGIN*/
class ParseTextIncident extends Effect {
  getScratchValue() {
    return "";
  }
  onProgress(fraction, currentTime) {
    for (const i in this.targetValue) {
      const { start, end, text } = this.targetValue[i].data;
      if (currentTime >= start && currentTime < end) {
        this.element.innerHTML = text;
        break;
      } else {
        this.element.innerHTML = "";
      }
    }
  }
}

const parseTextDefinition = {
  npm_name: "parse-text-definition",
  version: "1.0.0",
  incidents: [
    {
      exportable: ParseTextIncident,
      name: "ParseTextIncident",
      attributesValidationRules: {
        animatedAttrs: {
          type: "object",
          props: {
            subsArray: {
              type: "array",
            },
          },
        },
      },
    },
  ],
};

const ParseTextPlugin = loadPlugin(parseTextDefinition);

/*EXPORTED CLIP*/
export class ParseText extends HTMLClip {
  get html() {
    return `
    <div class="container">
      <div id="subs-container"></div>
    </div>
    `;
  }

  get css() {
    let position = "flex-end";

    if (this.attrs.position == "top") {
      position = "flex-start";
    } else if (this.attrs.position == "center") {
      position = "center";
    }

    return `
    .container {
      display:flex;
      justify-content:center;
      align-items:${position};
      font-size:${this.attrs.fontSize || 12}px;
      color:${this.attrs.textColor || "white"};
      font-family: ${this.attrs.fontFamily || "'Ubuntu'"};
      width: 100%;
      height: 100%;

    }
    #subs-container{
      max-width:${this.attrs.maxWidth ? this.attrs.maxWidth + "px" : "100%"};
      text-align:center;
      padding-top:${this.attrs.paddingTop || 0}px;
      padding-bottom:${this.attrs.paddingBottom || 0}px;
    }
  `;
  }

  buildTree() {
    try {
      const subs = parseSync(this.attrs.subtitles);
      const subtitle = new ParseTextPlugin.ParseTextIncident(
        {
          animatedAttrs: {
            subsArray: subs,
          },
        },
        {
          duration: subs[subs.length - 1].data.end,
          selector: "#subs-container",
        }
      );
      this.addIncident(subtitle, 0);
    } catch (e) {
      console.error("Error while loading subtitles", e);
    }
  }
}


demo/webpack.config.js
const path = require("path");
const webpack = require("webpack");

module.exports = {
  context: path.resolve(__dirname),

  entry: "./index.js",
  resolve: {
    fallback: {
      util: path.resolve(__dirname, "node_modules/util"),
      stream: require.resolve("stream-browserify"),
    },
  },
  output: {
    path: path.resolve(__dirname, "./"),
    // the output bundle
    filename: "./bundle.js",
  },

  module: {
    rules: [
      {
        test: /\.css$/,
        use: ["style-loader", "css-loader"],
      },
      {
        test: /\.js$/,
        use: "babel-loader",
        exclude: /node_modules/,
      },
    ],
  },

  plugins: [
    // enable HMR globally
    new webpack.HotModuleReplacementPlugin(),

    // do not emit compiled assets that include errors
    new webpack.NoEmitOnErrorsPlugin(),
    new webpack.DefinePlugin({
      "process.env": JSON.stringify(process.env),
    }),
  ],

  devServer: {
    host: "127.0.0.1",
    port: 8090,
    historyApiFallback: false,
    hot: false,
    static: path.join(__dirname),
  },
};


demo/index.js
import { HTMLClip, loadPlugin } from "@donkeyclip/motorcortex";
import Player from "@donkeyclip/motorcortex-player";
import SubtitlesDefinition from "../dist/motorcortex-subtitles.esm";
import { subs } from "./subs.srt.js";
const Subtitles = loadPlugin(SubtitlesDefinition);

const clip = new HTMLClip({
  html: `
    <div class="container">
      <div id="subs-container"></div>
    </div>`,
  css: `
  .container{
    width:100%;
    height:100%;
    position:relative;
    background:#151515;
  }
`,
  host: document.getElementById("clip"),
  containerParams: {
    width: "720px",
    height: "640px",
  },
});

const subtitle = new Subtitles.ParseText(
  {
    fontSize: 14,
    textColor: "white",
    fontFamily: "Ubuntu",
    subtitles: subs,
    position: "bottom",
    maxWidth: 400,
    paddingBottom: 50,
  },
  {
    selector: "#subs-container",
    containerParams: {
      width: "720px",
      height: "640px",
    },
  }
);

clip.addIncident(subtitle, 0);
new Player({
  scaleToFit: true,
  clip: clip,
  theme: "mc-blue",
  preview: true,
  pointerEvents: true,
});
